File: backend/src/api/v1/goals.py
"""
Goals API — /api/v1/goals

GET    /goals                          List goals
POST   /goals                          Create goal (with milestones)
GET    /goals/{id}                     Get goal
PATCH  /goals/{id}                     Update goal
DELETE /goals/{id}                     Delete goal
POST   /goals/{id}/milestones          Add milestone
PATCH  /goals/{id}/milestones/{ms_id}  Update milestone
DELETE /goals/{id}/milestones/{ms_id}  Delete milestone
"""
from uuid import UUID
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_

from src.core.dependencies import get_db, get_current_user
from src.schemas.goals import (
    GoalCreate, GoalUpdate, GoalRead, GoalListResponse,
    MilestoneCreate, MilestoneUpdate, MilestoneRead,
)
from src.services.goals_service import (
    create_goal, get_goal, list_goals, update_goal, delete_goal,
    add_milestone, update_milestone, delete_milestone,
)
from src.db.models.goal import Milestone
from src.core.logging import get_logger

router = APIRouter()
logger = get_logger(__name__)


def _not_found(label: str, uid: UUID) -> HTTPException:
    return HTTPException(
        404, detail={"detail": f"{label} {uid} not found", "code": f"{label.upper()}_NOT_FOUND"}
    )


@router.get("", response_model=GoalListResponse)
async def list_goals_endpoint(
    status: Optional[str] = Query(None, pattern="^(active|completed|paused|abandoned)$"),
    clerk_user_id: str = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> GoalListResponse:
    try:
        goals = await list_goals(db, clerk_user_id, status=status)
    except ValueError as e:
        raise HTTPException(404, detail={"detail": str(e), "code": "USER_NOT_FOUND"})
    items = [GoalRead.model_validate(g) for g in goals]
    return GoalListResponse(items=items, total=len(items))


@router.post("", response_model=GoalRead, status_code=201)
async def create_goal_endpoint(
    payload: GoalCreate,
    clerk_user_id: str = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> GoalRead:
    try:
        goal = await create_goal(db, clerk_user_id, payload)
    except ValueError as e:
        raise HTTPException(404, detail={"detail": str(e), "code": "USER_NOT_FOUND"})
    return GoalRead.model_validate(goal)


@router.get("/{goal_id}", response_model=GoalRead)
async def get_goal_endpoint(
    goal_id: UUID,
    clerk_user_id: str = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> GoalRead:
    goal = await get_goal(db, goal_id, clerk_user_id)
    if not goal:
        raise _not_found("Goal", goal_id)
    return GoalRead.model_validate(goal)


@router.patch("/{goal_id}", response_model=GoalRead)
async def update_goal_endpoint(
    goal_id: UUID,
    payload: GoalUpdate,
    clerk_user_id: str = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> GoalRead:
    goal = await get_goal(db, goal_id, clerk_user_id)
    if not goal:
        raise _not_found("Goal", goal_id)
    updated = await update_goal(db, goal, payload)
    return GoalRead.model_validate(updated)


@router.delete("/{goal_id}", status_code=204)
async def delete_goal_endpoint(
    goal_id: UUID,
    clerk_user_id: str = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    goal = await get_goal(db, goal_id, clerk_user_id)
    if not goal:
        raise _not_found("Goal", goal_id)
    await delete_goal(db, goal)


@router.post("/{goal_id}/milestones", response_model=GoalRead, status_code=201)
async def add_milestone_endpoint(
    goal_id: UUID,
    payload: MilestoneCreate,
    clerk_user_id: str = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> GoalRead:
    goal = await get_goal(db, goal_id, clerk_user_id)
    if not goal:
        raise _not_found("Goal", goal_id)
    updated = await add_milestone(db, goal, clerk_user_id, payload)
    return GoalRead.model_validate(updated)


@router.patch("/{goal_id}/milestones/{ms_id}", response_model=GoalRead)
async def update_milestone_endpoint(
    goal_id: UUID,
    ms_id:   UUID,
    payload: MilestoneUpdate,
    clerk_user_id: str = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> GoalRead:
    goal = await get_goal(db, goal_id, clerk_user_id)
    if not goal:
        raise _not_found("Goal", goal_id)
    ms_result = await db.execute(
        select(Milestone).where(and_(Milestone.id == ms_id, Milestone.goal_id == goal_id))
    )
    ms = ms_result.scalar_one_or_none()
    if not ms:
        raise _not_found("Milestone", ms_id)
    updated = await update_milestone(db, ms, goal, payload)
    return GoalRead.model_validate(updated)


@router.delete("/{goal_id}/milestones/{ms_id}", response_model=GoalRead)
async def delete_milestone_endpoint(
    goal_id: UUID,
    ms_id:   UUID,
    clerk_user_id: str = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> GoalRead:
    goal = await get_goal(db, goal_id, clerk_user_id)
    if not goal:
        raise _not_found("Goal", goal_id)
    ms_result = await db.execute(
        select(Milestone).where(and_(Milestone.id == ms_id, Milestone.goal_id == goal_id))
    )
    ms = ms_result.scalar_one_or_none()
    if not ms:
        raise _not_found("Milestone", ms_id)
    updated = await delete_milestone(db, ms, goal)
    return GoalRead.model_validate(updated)

File: frontend/src/lib/goals.ts
/**
 * Goals API service — all HTTP calls for the Goals module.
 */
import { apiClient } from "@/lib/api";
import type {
  Goal, GoalCreate, GoalUpdate, GoalListResponse,
  MilestoneCreate, MilestoneUpdate, GoalStatus,
} from "@/types/goals";

const BASE = "/api/v1/goals";

function mapMilestone(r: Record<string, unknown>) {
  return {
    id:         r.id as string,
    goalId:     r.goal_id as string,
    userId:     r.user_id as string,
    title:      r.title as string,
    isDone:     Boolean(r.is_done),
    dueDate:    (r.due_date as string | null) ?? null,
    orderIndex: (r.order_index as number) ?? 0,
    createdAt:  r.created_at as string,
  };
}

function mapGoal(r: Record<string, unknown>): Goal {
  return {
    id:          r.id as string,
    userId:      r.user_id as string,
    title:       r.title as string,
    description: (r.description as string | null) ?? null,
    status:      r.status as GoalStatus,
    targetDate:  (r.target_date as string | null) ?? null,
    color:       (r.color as string | null) ?? null,
    icon:        (r.icon as string | null) ?? null,
    progress:    (r.progress as number) ?? 0,
    createdAt:   r.created_at as string,
    updatedAt:   r.updated_at as string,
    milestones:  ((r.milestones ?? []) as Record<string, unknown>[]).map(mapMilestone),
  };
}

export async function fetchGoals(status?: GoalStatus): Promise<GoalListResponse> {
  const params: Record<string, string> = {};
  if (status) params.status = status;
  const res = await apiClient.get(BASE, { params });
  return {
    items: (res.data.items as Record<string, unknown>[]).map(mapGoal),
    total: res.data.total as number,
  };
}

export async function fetchGoal(id: string): Promise<Goal> {
  const res = await apiClient.get(`${BASE}/${id}`);
  return mapGoal(res.data as Record<string, unknown>);
}

export async function createGoal(payload: GoalCreate): Promise<Goal> {
  const body = {
    title:       payload.title,
    description: payload.description ?? null,
    status:      payload.status ?? "active",
    target_date: payload.targetDate ?? null,
    color:       payload.color ?? null,
    icon:        payload.icon ?? null,
    milestones:  (payload.milestones ?? []).map((m, i) => ({
      title: m.title, due_date: m.dueDate ?? null, order_index: i,
    })),
  };
  const res = await apiClient.post(BASE, body);
  return mapGoal(res.data as Record<string, unknown>);
}

export async function updateGoal(id: string, payload: GoalUpdate): Promise<Goal> {
  const body: Record<string, unknown> = {};
  if (payload.title       !== undefined) body.title       = payload.title;
  if (payload.description !== undefined) body.description = payload.description;
  if (payload.status      !== undefined) body.status      = payload.status;
  if (payload.targetDate  !== undefined) body.target_date = payload.targetDate;
  if (payload.color       !== undefined) body.color       = payload.color;
  if (payload.icon        !== undefined) body.icon        = payload.icon;
  if (payload.progress    !== undefined) body.progress    = payload.progress;
  const res = await apiClient.patch(`${BASE}/${id}`, body);
  return mapGoal(res.data as Record<string, unknown>);
}

export async function deleteGoal(id: string): Promise<void> {
  await apiClient.delete(`${BASE}/${id}`);
}

export async function addMilestone(goalId: string, payload: MilestoneCreate): Promise<Goal> {
  const body = {
    title: payload.title, due_date: payload.dueDate ?? null,
    order_index: payload.orderIndex ?? 0,
  };
  const res = await apiClient.post(`${BASE}/${goalId}/milestones`, body);
  return mapGoal(res.data as Record<string, unknown>);
}

export async function updateMilestone(
  goalId: string, msId: string, payload: MilestoneUpdate
): Promise<Goal> {
  const body: Record<string, unknown> = {};
  if (payload.title      !== undefined) body.title       = payload.title;
  if (payload.isDone     !== undefined) body.is_done     = payload.isDone;
  if (payload.dueDate    !== undefined) body.due_date    = payload.dueDate;
  if (payload.orderIndex !== undefined) body.order_index = payload.orderIndex;
  const res = await apiClient.patch(`${BASE}/${goalId}/milestones/${msId}`, body);
  return mapGoal(res.data as Record<string, unknown>);
}

export async function deleteMilestone(goalId: string, msId: string): Promise<Goal> {
  const res = await apiClient.delete(`${BASE}/${goalId}/milestones/${msId}`);
  return mapGoal(res.data as Record<string, unknown>);
}
